includes/wccom-site/class-wc-wccom-site-installer.php

```diff
@@ -0,0 +1,543 @@
+	 *
+	 * @return \WP_Error|string
+	 */
+	private static function download_product( $product_id, $upgrader ) {
+		$steps = self::get_state( 'steps' );
+		if ( empty( $steps[ $product_id ]['download_url'] ) ) {
+			return new WP_Error( 'missing_download_url', __( 'Could not find download url for the product.', 'woocommerce' ) );
+		}
+		return $upgrader->download_package( $steps[ $product_id ]['download_url'] );
+	}
+
+	/**
+	 * Unpack downloaded product.
+	 *
+	 * @since 3.7.0
+	 *
+	 * @param int          $product_id Product ID.
+	 * @param \WP_Upgrader $upgrader   Core class to handle installation.
+	 *
+	 * @return \WP_Error|string
+	 */
+	private static function unpack_product( $product_id, $upgrader ) {
+		$steps = self::get_state( 'steps' );
+		if ( empty( $steps[ $product_id ]['download_path'] ) ) {
+			return new WP_Error( 'missing_download_path', __( 'Could not find download path.', 'woocommerce' ) );
+		}
+
+		return $upgrader->unpack_package( $steps[ $product_id ]['download_path'], true );
+	}
+
+	/**
+	 * Move product to plugins directory.
+	 *
+	 * @since 3.7.0
+	 *
+	 * @param int          $product_id Product ID.
+	 * @param \WP_Upgrader $upgrader   Core class to handle installation.
+	 *
+	 * @return array|\WP_Error
+	 */
+	private static function move_product( $product_id, $upgrader ) {
+		$steps = self::get_state( 'steps' );
+		if ( empty( $steps[ $product_id ]['unpacked_path'] ) ) {
+			return new WP_Error( 'missing_unpacked_path', __( 'Could not find unpacked path.', 'woocommerce' ) );
+		}
+
+		$destination = 'plugin' === $steps[ $product_id ]['product_type']
+			? WP_PLUGIN_DIR
+			: get_theme_root();
+
+		$package = array(
+			'source'        => $steps[ $product_id ]['unpacked_path'],
+			'destination'   => $destination,
+			'clear_working' => true,
+			'hook_extra'    => array(
+				'type'   => $steps[ $product_id ]['product_type'],
+				'action' => 'install',
+			),
+		);
+
+		return $upgrader->install_package( $package );
+	}
+
+	/**
+	 * Activate product given its product ID.
+	 *
+	 * @since 3.7.0
+	 *
+	 * @param int $product_id Product ID.
+	 *
+	 * @return \WP_Error|null
+	 */
+	private static function activate_product( $product_id ) {
+		$steps = self::get_state( 'steps' );
+		if ( ! $steps[ $product_id ]['activate'] ) {
+			return null;
+		}
+
+		if ( 'plugin' === $steps[ $product_id ]['product_type'] ) {
+			return self::activate_plugin( $product_id );
+		}
+		return self::activate_theme( $product_id );
+	}
+
+	/**
+	 * Activate plugin given its product ID.
+	 *
+	 * @since 3.7.0
+	 *
+	 * @param int $product_id Product ID.
+	 *
+	 * @return \WP_Error|null
+	 */
+	private static function activate_plugin( $product_id ) {
+		// Clear plugins cache used in `WC_Helper::get_local_woo_plugins`.
+		wp_clean_plugins_cache();
+		$filename = false;
+
+		// If product is WP.org one, find out its filename.
+		$dir_name = self::get_wporg_product_dir_name( $product_id );
```

Comment:
For the future, something WCCOM could provide the client? This refers if it is hosted on wp.org or wc.com.

Reply:
Could also simplify function by handling the file retrieval for wp.org and wc.com in a separate function.

Reply:
Same applies to themes.

Reply:
> For the future, something WCCOM could provide the client?

We don't store main filename of extension in the product.

> Could also simplify function by handling the file retrieval for wp.org and wc.com in a separate function.

Given a product ID, for WCCOM products there's no special handling as helper caches plugin/theme info with valid subscriptions. However, it's not the case with wporg product in which WCCOM doesn't create subscription, thus no lookup provided.